### ISSUE TITLE
How to use softlabel in MXNET?

### ISSUE THREAD

**Issue by xuliuac** [apache/incubator-mxnet#8025]
Hi,
Does any one know how to use softlabel such as [0.7,0,0,0.3,0,0,0],rather than one hot label[1,0,0,0,0,0] in image classification task? I need set soft label for distilling the dark knowledge during training. But I have no ideal how to do it?

Many thanks!

**Comment by Ldpe2G** [comment]
try L2 Loss

**Comment by unrealwill** [comment]
You probably can make your own loss function
```
label =  mx.symbol.Variable('label')  
mlp = mx.symbol.FullyConnected(...)
pred = mx.symbol.softmax( mlp ,axis=1)
loss = mx.symbol.MakeLoss( mx.symbol.sum( - mx.symbol.log( pred ) *label ,axis=1) ) 
```

make sure to pass a label variable which sums to one.
I'm sure you can optimize it with the logsoftmax operators.

**Comment by xuliuac** [comment]
@unrealwill 
hi,thanks for your answer.I have try your code but it does not work in my case ,my net not converge by this loss,Now I am still checking where is wrong in my code .

But I have found another solution to deal with softlabel. which is simply pass the softlabel to mx.symbol.SoftmaxOutput() and write a custom metric for it. I am new in mxnet, if I do something wrong,please correct me. 
Thanks again, cheers.

**Comment by unrealwill** [comment]
@xuliuac 
I'm not quite familiar with mxnet yet, but I think that metric and losses are two different things. If mxnet is anything like Keras, Metrics are being used for reporting only (additional info), and loss being used for actual training.

(If I didn't make a sign error) The loss won't converge to 0 (you can calculate the optimum value the loss could attain : (sum(-log(max(label)) )) (a soft label is similar to training set error (two training examples with different label) like saying here is this example I don't know if it's a 3 or a 8 but I'm 65% confident it's a 8 ), but it should converge towards a network trying to reach certainty in its prediction (a deterministic policy) (It will predict a 8 100%, but not get penalized too much if he predict a 3 ).

If you try to learn a stochastic policy then use a Kullback-Leibler divergence between you prediction and your soft label.

**Comment by xuliuac** [comment]
@unrealwill 
Yes, as you mentioned above,metrics has nothing to do with training. It seems like the  mx.symbol.SoftmaxOutput() is support softlabel already and I just use a inappropriate eval metric at beginning.  

After check the definition ,I think I got your point about using Kullback-Leibler divergence,my task is about video content classification,some video may have two label,such like 'dance' and 'sing', prediction is equally right for both label,so I think in this case cross entropy is good enough for the training.Now I have run the training task for a few hours, it do show the training accuracy increased much faster than use onehot label. I hope the final test accuracy is better than before. (๑•̀ㅂ•́)و✧

**Comment by unrealwill** [comment]
@xuliuac 
Also if your can have multiple label per video, then don't use a softmax as the last layer but rather a sigmoid with a binary cross-entropy loss. And not use a softlabel which sums to 1.0 but rather a label which is 1.0 for each feature which should be on.

It depends on whether or not your categories are mutually exclusive or not.

**Comment by chinakook** [comment]
Your label may represent attributes, so you can try **logistic loss** on multiple output channels in which each output channel represent prob of an attribute.
